File: data structures and algorithms/cpp/08_048.cpp
#include <vector>
#include <queue>
#include <cstddef> // for nullptr

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* interleaveLists(
            std::vector<ListNode*>& lists) {
        if (lists.empty()) {
            return nullptr;
        }
        
        // Create a queue to store non-empty lists
        std::queue<ListNode*> queue;
        for (ListNode* list : lists) {
            if (list != nullptr) {
                queue.push(list);
            }
        }
        
        if (queue.empty()) {
            return nullptr;
        }
        
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (!queue.empty()) {
            // Get current list
            ListNode* current = queue.front();
            queue.pop();
            
            // Append current node to result
            tail->next = current;
            tail = tail->next;
            
            // Move to next node in current list
            if (current->next != nullptr) {
                queue.push(current->next);
            }
        }
        
        return dummy.next;
    }
};

File: data structures and algorithms/java/08_048.java
class ListNode {
    int val;
    ListNode next;
    
    ListNode(int val) {
        this.val = val;
    }
}

public class Solution {
    public ListNode interleaveLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        
        // Create a new list to store non-empty lists
        List<ListNode> activeList = new ArrayList<>();
        for (ListNode list : lists) {
            if (list != null) {
                activeList.add(list);
            }
        }
        
        if (activeList.isEmpty()) {
            return null;
        }
        
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;
        
        while (!activeList.isEmpty()) {
            // Process one node from each list
            int size = activeList.size();
            for (int i = 0; i < size; i++) {
                ListNode current = activeList.remove(0);
                
                // Append current node to result
                tail.next = current;
                tail = tail.next;
                
                // Move to next node in current list
                if (current.next != null) {
                    activeList.add(current.next);
                }
            }
        }
        
        return dummy.next;
    }
}